src/Compilers/Core/Portable/Syntax/SyntaxNode.cs

```diff
@@ -611,25 +611,52 @@ internal int GetChildIndex(int slot)
         /// efficient implementations.
         /// </summary>
         internal virtual int GetChildPosition(int index)
+        {
+            return GetChildPosition(index, useNextNotPrevious: false);
+        }
+
+        internal int GetChildPosition(int index, bool useNextNotPrevious)
         {
             int offset = 0;
             var green = this.Green;
-            while (index > 0)
+
+            if (!useNextNotPrevious)
             {
-                index--;
-                var prevSibling = this.GetCachedSlot(index);
-                if (prevSibling != null)
+                while (index > 0)
                 {
-                    return prevSibling.EndPosition + offset;
+                    index--;
+                    var prevSibling = this.GetCachedSlot(index);
+                    if (prevSibling != null)
+                    {
+                        return prevSibling.EndPosition + offset;
+                    }
+                    var greenChild = green.GetSlot(index);
+                    if (greenChild != null)
+                    {
+                        offset += greenChild.FullWidth;
+                    }
                 }
-                var greenChild = green.GetSlot(index);
-                if (greenChild != null)
+                return this.Position + offset;
+            }
+            else
+            {
+                int slotCount = green.SlotCount;
+                while (index < slotCount - 1)
                 {
-                    offset += greenChild.FullWidth;
+                    index++;
+                    var prevSibling = this.GetCachedSlot(index);
+                    if (prevSibling != null)
+                    {
+                        return prevSibling.EndPosition - offset;
+                    }
+                    var greenChild = green.GetSlot(index);
+                    if (greenChild != null)
+                    {
+                        offset += greenChild.FullWidth;
+                    }
                 }
+                return this.EndPosition - offset;
             }
-
-            return this.Position + offset;
         }
```

Comment:
overall, i have no issue with the approach.  but i think we *def* need tests around ensuring that reverse iterating produces nodes/tokens in the correct location.  can you add a bunch of tests that show that if we reverse iterate the positions of things are all correct?

Reply:
Added verifying results from `GetChildPosition()` and `GetChildPositionFromEnd()` to tests in IOperation verification.